pygments/lexers/ul4.py

```diff
@@ -0,0 +1,261 @@
+                # ``<?ul4?>``
+                r"(<\?)(\s*)(ul4)(\s*)(\?>)",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword, token.Text, token.Comment.Preproc),
+            ),
+            (
+                # Template header with name (potentially followed by the signature):
+                # ``<?ul4 foo(bar=42)?>``
+                r"(<\?)(\s*)(ul4)(\s*)([a-zA-Z_][a-zA-Z_0-9]*)?\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword, token.Text, token.Token.Name.Function),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # Comment:
+                # ``<?note foobar?>``
+                r"<\?\s*note\s+.*?\?>",
+                token.Comment,
+            ),
+            (
+                # Template documentation:
+                # ``<?doc foobar?>``
+                r"<\?\s*doc\s+.*?\?>",
+                token.String.Doc,
+            ),
+            (
+                # ``<?ignore?>`` tag for commenting out code:
+                # ``<?ignore?>...<?end ignore?>``
+                r"<\?\s*ignore\s*\?>",
+                token.Comment,
+                "ignore", # Switch to "ignore" mode
+            ),
+            (
+                # ``<?def?>`` tag for defining local templates
+                # ``<?def foo(bar=42)?>...<?end def?>``
+                r"(<\?\s*)(def)(\s*)([a-zA-Z_][a-zA-Z_0-9]*)?\b",
+                lexer.bygroups(token.Comment.Preproc, token.Keyword, token.Text, token.Token.Name.Function),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # The rest of the supported tags
+                r"(<\?)(\s*)(printx|print|for|if|elif|else|while|code|renderblocks?|render)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # ``<?end?>`` tag for ending ``<?def?>``, ``<?for?>``,
+                # ``<?if?>``, ``<?while?>``, ``<?renderblock?>`` and
+                # ``<?renderblocks?>`` blocks.
+                r"(<\?)(\s*)(end)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "end", # Switch to "end tag" mode
+            ),
+            (
+                # ``<?whitespace?>`` tag for configuring whitespace handlng
+                r"(<\?)(\s*)(whitespace)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "whitespace", # Switch to "whitespace" mode
+            ),
+            # Plain text
+            (r"[^<]+", token.Token.Other),
+            (r"<", token.Token.Other),
+        ],
+        # Ignore mode ignores everything upto the matching ``<?end ignore?>`` tag
+        "ignore": [
+            # Nested ``<?ignore?>`` tag
+            (r"<\?\s*ignore\s*\?>", token.Comment, "ignore"),
```

Comment:
For symmetry I'd do

```suggestion
            (r"<\?\s*ignore\s*\?>", token.Comment, "#push"),
```

Reply:
Fixed in https://github.com/doerwalter/pygments/commit/6ca0fae5d07ad1b29908f0cf067a3f87f98afa47